src/PooledArrays.jl

```diff
@@ -439,40 +446,49 @@ Base.convert(::Type{Array}, pa::PooledArray{T, R, N}) where {T, R, N} = convert(
 
-# Scalar case
-Base.@propagate_inbounds function Base.getindex(pav::PooledArrOrSub, I::Integer...)
-    idx = DataAPI.refarray(pav)[I...]
-    iszero(idx) && throw(UndefRefError())
-    return @inbounds DataAPI.refpool(pav)[idx]
+# We need separate functions due to dispatch ambiguities
+
+for T in (PooledArray, SubArray{<:Any, <:Any, <:PooledArray})
+    @eval Base.@propagate_inbounds function Base.getindex(A::$T, I::Integer...)
+        idx = DataAPI.refarray(A)[I...]
+        iszero(idx) && throw(UndefRefError())
+        return @inbounds DataAPI.refpool(A)[idx]
+    end
+
+    @eval Base.@propagate_inbounds function Base.getindex(A::$T, I::Union{Real, AbstractVector}...)
+        # make sure we do not increase A.refcount in case creation of newrefs fails
+        newrefs = getindex(DataAPI.refarray(A), I...)
+        @assert newrefs isa AbstractArray
+        Threads.atomic_add!(refcount(A), 1)
+        return PooledArray(RefArray(newrefs), DataAPI.invrefpool(A), DataAPI.refpool(A), refcount(A))
+    end
 end
 
-# this is needed due to dispatch ambiguities
-Base.@propagate_inbounds function Base.getindex(pav::PooledArrOrSub{T,N}, I::Vararg{Int,N}) where {T,N}
-    idx = DataAPI.refarray(pav)[I...]
-    iszero(idx) && throw(UndefRefError())
-    return @inbounds DataAPI.refpool(pav)[idx]
+if VERSION < v"1.1"
+    Base.@propagate_inbounds function Base.getindex(A::SubArray{T,D,P,I,true} ,
+                                                    i::Int) where {I<:Tuple{Union{Base.Slice,
+                                                                                  AbstractUnitRange},
+                                                                            Vararg{Any}}, P<:PooledArray, T, D}
+        idx = DataAPI.refarray(A)[i]
+        iszero(idx) && throw(UndefRefError())
+        return @inbounds DataAPI.refpool(A)[idx]
+    end
 end
 
-Base.@propagate_inbounds function Base.isassigned(pa::PooledArray, I::Int...)
-    !iszero(pa.refs[I...])
+Base.@propagate_inbounds function Base.getindex(A::PooledArray{<:Any, <:Integer, N}, I::Vararg{Int,N}) where {T,N}
+    idx = DataAPI.refarray(A)[I...]
+    iszero(idx) && throw(UndefRefError())
+    return @inbounds DataAPI.refpool(A)[idx]
 end
 
-# Other cases; we rely on the fact that non-standard indexing will fall back to this eventually
-Base.@propagate_inbounds function Base.getindex(A::PooledArray, I::Union{Real, AbstractVector}...)
-    # make sure we do not increase A.refcount in case creation of newrefs fails
-    newrefs = getindex(A.refs, I...)
-    @assert newrefs isa AbstractArray
-    Threads.atomic_add!(A.refcount, 1)
-    return PooledArray(RefArray(newrefs), A.invpool, A.pool, A.refcount)
+Base.@propagate_inbounds function Base.getindex(A::SubArray{<:Any, N, <:PooledArray}, I::Vararg{Int,N}) where {T,N}
+    idx = DataAPI.refarray(A)[I...]
+    iszero(idx) && throw(UndefRefError())
+    return @inbounds DataAPI.refpool(A)[idx]
 end
```

Comment:
Also put these in the loop above? Also isn't this redundant with the `I::Int` method if we require `IndexLinear`?

Reply:
I cannot put them in the loop above because parameter `N` has to be matched.

Also we have to define them unfortunately, as otherwise there are dispatch ambiguities. The problem is that even if we are `IndexLinear` we have to define a special case `I::Union{Real, AbstractVector}...` above and `I::Vararg{Int,N}` also matches this pattern and `I::Vararg{Int,N}` is defined in Julia Base as a special case for `SubArray`.

But I could remove it for `PooledArray`.